osbuild: support building multiple platforms at once

dustymabe opened the pull request:
This is done by adding `cosa osbuild qemu metal...`.

There is a lot of refactor/cleanup in this PR. Please review commits individually.

ravanelli commented on src/cmd-buildextend-metal:
```diff
@@ -139,24 +139,37 @@ fi
+}
+
+# Grab a few values from ${image_json}
+deploy_via_container=$(getconfig_def "deploy-via-container" "" "${image_json}")
```
We are removing the `deploy_via_container=""` what will be the return if we don't have it in the image_json, it maybe break the podman build

dustymabe replied:
```
getconfig_def "deploy-via-container" "" "${image_json}"
```

The function sets a default value to `""` (which we are passing here as the 2nd argument) if there is no value for it in the image.json.

jlebon commented on src/cmd-buildextend-metal:
```diff
@@ -31,6 +31,32 @@ getconfig_def() {
+    'path': '${target_filename}',
+    'sha256': '$(sha256sum_str < "${local_filepath}")',
+    'size': $(stat -c '%s' "${local_filepath}"),
+    'skip-compression': ${skip_compress}
```
Minor: this is mostly cosmetic, but let's only add the key if it's `true` to retain the current behaviour? That makes it easier to tell when scanning the `meta.json` what artifacts have it.

jlebon commented on src/cmd-osbuild:
```diff
@@ -8,10 +8,10 @@ dn=$(dirname "$0")
-Usage: coreos-assembler buildextend-${platform} --help
-       coreos-assembler buildextend-${platform} [--build ID]
+Usage: coreos-assembler osbuild --help
+       coreos-assembler osbuild [--build ID] --platforms qemu,metal,...
```
Hmm, WDYT about:

```
cosa osbuild qemu metal ...
```

?

Seems more ergonomic and matches `cosa build`.

dustymabe replied:
OK. Will update to do it that way.

jlebon commented on src/cmd-osbuild:
```diff
@@ -267,66 +273,84 @@ main() {
+    for platform in "${platforms[@]}"; do
+        suffix=
+        case $platform in
+            applehv|metal|metal4k) suffix=raw ;;
+            gcp) suffix=tar.gz ;;
+            hyperv) suffix=vhdx ;;
+            qemu|qemu-secex) suffix=qcow2 ;;
```
I'm confused why we need this. Doesn't osbuild already output the files with the right extension?

dustymabe replied:
Previously OSBuild output to whatever filename we told it to. It was a config parameter that was passed in, but that really only worked in the 1:1 coupling we had where we ran OSBuild separately for each platform. In this PR I change the final pipelines for each platform to just output to a filename that matches the platform name (i.e. no suffix).

We could hardcode the suffix inside the OSBuild manifest yaml and maybe we should, but we'd still need to know the suffix here in `cmd-osbuild` in order to copy the file out (i.e. we'd copy `$outdir/$platform/$platform.$suffix` out instead of `$outdir/$platform/$platform`).

So either way this script needs to know what the suffix will be. Whether we add the suffix to the intermediate file generated by OSBuild in the outdir isn't important to me.

jlebon replied:
Right, got it. Hmm, it does seem to me like osbuild should be outputting files with the right extension to make it more self-contained and usable outside of cosa. You're right though that it wouldn't change the fact that this code would also need to know about extensions (unless we use globs I guess, but that feels yucky). Anyway, no need to block on this.

dustymabe replied:
> Right, got it. Hmm, it does seem to me like osbuild should be outputting files with the right extension to make it more self-contained and usable outside of cosa.

OK. I found a way to have the actual file names OSBuild generates in the outdir to be the correct filenames we want to store in the builddir. Updated.

jlebon reviewed (approved):
Very nice! Overall LGTM.

dustymabe commented:
Pushed an update to address code review comments and also update a few missing things from the gcp/applehv/hyperv platforms.

dustymabe commented:
ideas on this shellcheck error?

```
./tests/check_one.sh src/cmd-osbuild src/.cmd-osbuild.shellchecked
if ${skip_compress}:
Possible unbracketed conditional in src/cmd-osbuild
make: *** [Makefile:39: src/.cmd-osbuild.shellchecked] Error 1
```

The code it's complaining about is embedded python.

dustymabe commented:
> ideas on this shellcheck error?
> 
> ```
> ./tests/check_one.sh src/cmd-osbuild src/.cmd-osbuild.shellchecked
> if ${skip_compress}:
> Possible unbracketed conditional in src/cmd-osbuild
> make: *** [Makefile:39: src/.cmd-osbuild.shellchecked] Error 1
> ```
> 

It's not actually shellcheck that is complaining. It's hitting this check that we added in https://github.com/coreos/coreos-assembler/pull/3155 

https://github.com/coreos/coreos-assembler/blob/48fba72a62c739f6eaf4d7991052516fd574acc7/tests/check_one.sh#L30-L33

dustymabe commented:
> It's not actually shellcheck that is complaining. It's hitting this check that we added in #3155

OK. I was able to figure out a workaround.. rebased the PR and added a commit with the workaround.

jlebon commented on src/cmd-osbuild:
```diff
@@ -0,0 +1,391 @@
+    'sha256': '$(sha256sum_str < "${local_filepath}")',
+    'size': $(stat -c '%s' "${local_filepath}")
+}
+# backwards conditional here because of a invalid (for this code) CI check
+# https://github.com/coreos/coreos-assembler/pull/3930#issuecomment-2473714222
+if False:
+    pass
+elif ${skip_compress}:
+    j['images']['${artifact_name}']['skip-compression'] = True
```
Heh, that's actually kinda funny.

I think another way around this is:

```
if True == ${skip_compression}:
    ...
```

Not worth the respin

jlebon reviewed (approved):
LGTM!
